RandLAPACK/comps/rl_util.hh

```diff
@@ -552,5 +813,47 @@ int64_t rank_search_binary(
     return k;
 }
 
+/// Iterates the upper-triangular matrix A from end to beginning and 
+/// attempts to estimate smallest k such taht ||A[k:, k:]||_F <= tau_trunk * ||A||_2.
+/// Computes matrix Fro norms through vector norm updates.
+template <typename T>
+int64_t rank_search_linear(
+    int64_t n,
+    T norm_2_A,
+    T tau_trunc,
+    T* A_dat
+) {
+    T norm_A_work = 0.0;
+    T norm_A_row = 0.0;
+    for(int i = n - 1, j = 1; i > 0; --i, ++j) {
+        norm_A_row = blas::nrm2(j, A_dat + ((n + 1) * i), n);
+        // Add norm stably
+        norm_A_work = std::hypot(norm_A_work, norm_A_row);
+
+        if(norm_A_work > tau_trunc * norm_2_A)
+            return i + 1;
+    }
+    return 1;
+}
+
+/// Normalizes columns of a given matrix, writes the result into a buffer
+template <typename T>
+void normc(
+    int64_t m,
+    int64_t n,
+    std::vector<T>& A,
+    std::vector<T>& A_norm
+) {
+    util::upsize(m * n, A_norm);
+
+    T col_nrm = 0.0;
+    for(int i = 0; i < n; ++i) {
+        col_nrm = blas::nrm2(m, &A[m * i], 1);
+        for (int j = 0; j < m; ++j) {
+            A_norm[m * i + j] = A[m * i + j] / col_nrm;
+        }
```

Comment:
If ``A`` is column-major then you can just set ``inv_col_nrm = 1/col_nrm`` and then use ``blas::scal``.

Reply:
Will that actually be faster?

Reply:
No, but it will be more readable.

Reply:
*More readable for people who are used to reading linear algebra code. There's a "reasonable" expectation that if a BLAS function exists for some operation, then you're going to use it. Framed another way, if people see a for loop that operates on the entries of a matrix they will assume that the purpose of the for loop is something that can't be done with one BLAS call.

Reply:
Well, we're iterating through all entries in a given column-vector anyway. Why not scale those immediately instead of separately scaling the column vector after we've already assigned all values to it?

Reply:
@TeachRaccooon I actually missed the fact that ``A_norm`` was separate from ``A``.  It's obvious now.

You could do away with the second for-loop if ``A_norm`` started as all zeros. In that case, you could ``axpy`` each column of ``A`` into ``A_norm`` with a scaling factor ``1/col_nrm``. But since you don't have an assurance that ``A_norm`` is all zeros, your current implementation is fine.

One request: please go through these utility functions and mark values as ``const`` if the function isn't supposed to change them. In this case, I think you'd want the third argument to the function to be ``const std::vector<T>& A``. Try that out, and if it works, then please apply changes throughout ``rl_util.hh``.

Reply:
Done.
I actually did that before.
This function was the only place that needed change.